README
HTML::Examiner is a module to run rule based verifications on HTML. This 
is not a module to execute W3C validations. The rules are described using 
a Domain Specific Language. See module's documentation for more info.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc HTML::Examiner

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTML-Examiner

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/HTML-Examiner

    CPAN Ratings
        http://cpanratings.perl.org/d/HTML-Examiner

    Search CPAN
        http://search.cpan.org/dist/HTML-Examiner/


LICENSE AND COPYRIGHT

Copyright (C) 2010 Nuno Carvalho

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for HTML-Examiner

0.01_2 Thu May  6 22:09:11 WEST 2010
       Fix Makefile.PL dependencias.

0.01_1 Mon May  3 22:27:21 WEST 2010
       First beta version.



Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

# yapp -m HTML::Examiner::Parser -o lib/HTML/Examiner/Parser.pm lib/HTML/Examiner/Parser.yp

my $MY_YAPP = `which yapp`;
chomp $MY_YAPP;
$MY_YAPP eq '' and die "install Parse::Yapp";

`$MY_YAPP -m HTML::Examiner::Parser -o lib/HTML/Examiner/Parser.pm lib/HTML/Examiner/Parser.yp`;
if ($? == 0) { print "Building lib/HTML/Examiner/Parser.pm\n"; }
else { print "Building lib/HTML/Examiner/Parser.pm (failed)\n"; }

WriteMakefile(
    NAME                => 'HTML::Examiner',
    AUTHOR              => q{Nuno Carvalho <smash@cpan.org>},
    VERSION_FROM        => 'lib/HTML/Examiner.pm',
    ABSTRACT_FROM       => 'lib/HTML/Examiner.pm',
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More' => 0,
        'Template' => 0,
        'XML::DT' => 0,
        'Parse::Yapp' => 0,
        'Parse::Yapp::Driver' => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'HTML-Examiner-*' },
);


lib/HTML/Examiner.pm
package HTML::Examiner;

use warnings;
use strict;

use XML::DT;
use HTML::Examiner::Parser;
use HTML::Examiner::DTEngine;
use HTML::Examiner::Templates;

=head1 NAME

HTML::Examiner - verify HTML code using custom defined rules

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01_2';

my $report;

=head1 SYNOPSIS

  use HTML::Examiner;

  my $errors = {
    0 => {
      source => 'tag(img) misses prop(alt) => error "img misses tag";'
    }
  };

  my $examiner = HTML::Examiner->new($errors);
  my $report = $examiner->process($html);

=head1 DESCRIPTION

This module implements an engine to verify HTML code use custom
defined rules. This rules are defined in a Domain Specific Language.
For more documentation on writing rules see HTML::Examiner::Parser
documentation.

This module is *not* intended to perform traditional HTML validations,
use other modules for that purpose. Typically you don't even need
to have a valid HTML document, as long as it can be correctly
parsed by XML::DT.

=head1 METHODS

=head2 new

Create a new examiner instance. Reveices as argument an hash
representing our costum defined error database.

=cut

sub new {
	my($class,$errors) = @_;
	my $self = {};
	$self->{'errors'} = $errors;

	foreach (sort keys %$errors) {
		my $ptree = __parse_rule($self->{'errors'}->{$_}->{'source'});
		$self->{'errors'}->{$_}->{'ptree'} = $ptree;
	}

	bless $self, $class;
}

=head2 examine

Call this method to examine HTML code. The code is passed as an argument.

=cut

sub examine {
	my($self, $html) = @_;

	$report = {};
	foreach (sort keys %{$self->{'errors'}}) {
		my $engine = HTML::Examiner::DTEngine->new($self->{'errors'}->{$_}->{'ptree'});
		$engine->process($html);
	}

	HTML::Examiner::report();
}

=head2 __parse_rule

Method to parse a rule a custom defined rule. You shoulnd't have to
call this method directly.

=cut

sub __parse_rule {
	my $source = shift;

	my $parser = HTML::Examiner::Parser->new;
	$parser->init_lex($source);
	my $ptree = $parser->YYParse(
			yylex => \&HTML::Examiner::Parser::lex,
			yyerror => \&HTML::Examiner::Parser::yyerror
		);

	$ptree;
}

=head2 report

This method is used to feed report information when verifying
custom defined rules.

=cut

sub report {
	my $new = shift;

	defined($new) and $report->{$new}++;

	return [keys %$report];
}

=head1 AUTHOR

Nuno Carvalho, C<< <smash at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-html-examiner at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=HTML-Examiner>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc HTML::Examiner


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTML-Examiner>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/HTML-Examiner>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/HTML-Examiner>

=item * Search CPAN

L<http://search.cpan.org/dist/HTML-Examiner/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2010 Nuno Carvalho.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of HTML::Examiner


lib/HTML/Examiner/DTEngine.pm
package HTML::Examiner::DTEngine;

use warnings;
use strict;

use XML::DT;
use Template;
use HTML::Examiner::Templates;
use Template::Constants qw( :debug );
use Data::Dumper;

=head1 NAME

HTML::Examiner::DTEngine - The great new HTML::Examiner::DTEngine

=head1 VERSION

Version 0.01_1

=cut

our $VERSION = '0.01_2';

=head1 SYNOPSIS

  use HTML::Examiner::DTEngine;

  my $examiner = HTML::Examiner->new();

=head1 SUBROUTINES/METHODS

=head2 new

Create new module instance.

=cut

sub new {
	my($class,$ptree) = @_;

	my $self = {};
	__proc_rule($self,$ptree);

	bless $self, $class;
}

=head2 process

Process the HTML given as argument with the user defined rules.

=cut

sub process {
	my($self,$html) = @_;

	my %handler = (
		-html => 1,
	);
	$handler{$self->{'tag'}} = eval "sub {".$self->{'code'}." }";

	dtstring($html, %handler);
}

=head2 __proc_rule

Process a single rule, you shouldn't call this method directly.

=cut

sub __proc_rule {
	my($self,$ptree) = @_;

	# handle tag any
	if ($ptree->{'function'}->{'pattern'}->{'target'}->{'tag'}->{'name'} eq 'ANY' or defined($ptree->{'function'}->{'pattern'}->{'target'}->{'tag'}->{'re'}) ) {
		$self->{'tag'} = '-default';
	}
	else {
		$self->{'tag'} = $ptree->{'function'}->{'pattern'}->{'target'}->{'tag'}->{'name'};
	}

	my $template_config = {
			INCLUDE_PATH => [ 'templates' ],
		};

	my $template = Template->new({
        LOAD_TEMPLATES => [ HTML::Examiner::Templates->new($template_config) ],
	});

	my $output;
	my $vars = {};
	$vars->{'ptree'} = $ptree;

	# handle target for template name
	my $template_name = 'tag';
	if (defined($ptree->{'function'}->{'pattern'}->{'target'}->{'prop'})) {
		$template_name = 'prop';
	}
	if (defined($ptree->{'function'}->{'pattern'}->{'target'}->{'content'})) {
		$template_name = 'content';
	}
	if (defined($ptree->{'function'}->{'pattern'}->{'target'}->{'tag'}->{'re'})) {
		$template_name = 're';
	}
	if ($ptree->{'function'}->{'pattern'}->{'target'}->{'tag'}->{'name'} eq 'ANY') {
		$template_name = 'any';
	}

	# handle verb template name
	if (defined($ptree->{'function'}->{'pattern'}->{'verb'}->{'verb'})) {
		$template_name .= '_'.$ptree->{'function'}->{'pattern'}->{'verb'}->{'verb'};
	}
	else {
		$template_name .= '_exists';
	}

	# handle last part of template name
	if (defined($ptree->{'function'}->{'pattern'}->{'element'}->{'prop'})) {
		$template_name .= '_prop';
	}
	if (defined($ptree->{'function'}->{'pattern'}->{'element'}->{'re'})) {
		$template_name .= '_re';
	}

	# process template
	$template->process($template_name, $vars, \$output);
	$self->{'code'} = $output;
}

=head1 AUTHOR

Nuno Carvalho, C<< <smash at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-html-examiner at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=HTML-Examiner>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc HTML::Examiner


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTML-Examiner>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/HTML-Examiner>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/HTML-Examiner>

=item * Search CPAN

L<http://search.cpan.org/dist/HTML-Examiner/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2010 Nuno Carvalho.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of HTML::Examiner


lib/HTML/Examiner/Templates.pm
package HTML::Examiner::Templates;
use base qw( Template::Provider );

use warnings;
use strict;

=head1 NAME

HTML::Examiner::Templates - this module store the templates

=head1 VERSION

Version 0.01_1

=cut

our $VERSION = '0.01_2';

my $templates = {
'tag_contains_prop'=> <<'EOT'
if (defined($v{[% ptree.function.pattern.element.prop %]})) { 
    [% PROCESS src/error_report msg = ptree.function.action.msg %]
}
EOT
,
'src/error_report'=> <<'EOT'
	HTML::Examiner::report("[% msg %]");
EOT
,
'any_contains_prop' => <<'EOT'
if (defined($v{[% ptree.function.pattern.element.prop %]})) { 
    [% PROCESS src/error_report msg = ptree.function.action.msg %]
}
EOT
,
'any_misses_prop' => <<'EOT'
unless (defined($v{[% ptree.function.pattern.element.prop %]})) { 
    [% PROCESS src/error_report msg = ptree.function.action.msg %]
}
EOT
,
'content_matches_re' => <<'EOT'
if ($c =~ /[% ptree.function.pattern.element.re %]/) {
    [% PROCESS src/error_report msg = ptree.function.action.msg %]
}
EOT
,
'content_mismatches_re' => <<'EOT'
if (not($c =~ /[% ptree.function.pattern.element.re %]/)) {
    [% PROCESS src/error_report msg = ptree.function.action.msg %]
}
EOT
,
'prop_matches_re' => <<'EOT'
if (defined($v{[% ptree.function.pattern.target.prop %]}) and ($v{[% ptree.function.pattern.target.prop %]} =~ /[% ptree.function.pattern.element.re %]/)) { 
    [% PROCESS src/error_report msg = ptree.function.action.msg %]
}
EOT
,
'prop_mismatches_re' => <<'EOT'
if (defined($v{[% ptree.function.pattern.target.prop %]}) and not($v{[% ptree.function.pattern.target.prop %]} =~ /[% ptree.function.pattern.element.re %]/)) { 
    [% PROCESS src/error_report msg = ptree.function.action.msg %]
}
EOT
,
're_contains_prop' => <<'EOT'
if ($q =~ /[% ptree.function.pattern.target.tag.re %]/ and defined($v{[% ptree.function.pattern.element.prop %]})) { 
	push @{$HTML::Examiner::report}, "[% ptree.function.action.msg %]";
}
EOT
,
're_misses_prop' => <<'EOT'
unless ($q =~ /[% ptree.function.pattern.target.tag.re %]/ and defined($v{[% ptree.function.pattern.element.prop %]})) { 
	push @{$HTML::Examiner::report}, "[% ptree.function.action.msg %]";
}
EOT
,
'tag_exists' => <<'EOT'
[% PROCESS src/error_report msg = ptree.function.action.msg %]
EOT
,
'tag_misses_prop' => <<'EOT'
unless (defined($v{[% ptree.function.pattern.element.prop %]})) { 
	[% PROCESS src/error_report msg = ptree.function.action.msg %]
}
EOT
};

sub _template_modified {
    my($self,$path) = @_;
	
	return 1;
}

sub _template_content {
    my($self,$path) = @_;

	$path =~ s#^templates/##;
    $self->debug("get $path") if $self->{DEBUG};
	
	my $data = $templates->{$path};
	my $error = "error: $path not found";
	my $mod_date = 1;

	return $data;
}

1;


lib/HTML/Examiner/Parser.pm
####################################################################
#
#    This file was generated using Parse::Yapp version 1.05.
#
#        Don't edit this file, use source file instead.
#
#             ANY CHANGE MADE HERE WILL BE LOST !
#
####################################################################
package HTML::Examiner::Parser;
use vars qw ( @ISA );
use strict;

@ISA= qw ( Parse::Yapp::Driver );
use Parse::Yapp::Driver;

#line 1 "lib/HTML/Examiner/Parser.yp"

=head1 NAME

HTML::Examiner::Parser - this module implements the parser

=head1 VERSION

Version 0.01_1

=cut

our $VERSION = '0.01_1';

=head1 DESCRIPTION

XXX

=head1 FUNCTIONS

=cut

=head2 new

Create a new parser.

=cut


sub new {
        my($class)=shift;
        ref($class)
    and $class=ref($class);

    my($self)=$class->SUPER::new( yyversion => '1.05',
                                  yystates =>
[
	{#State 0
		ACTIONS => {
			'TAG' => 3
		},
		GOTOS => {
			'pattern' => 2,
			'target' => 1,
			'function' => 5,
			'statement' => 4
		}
	},
	{#State 1
		ACTIONS => {
			'VERB' => 6
		},
		DEFAULT => -4,
		GOTOS => {
			'verb' => 7
		}
	},
	{#State 2
		ACTIONS => {
			'ARROW' => 8
		}
	},
	{#State 3
		ACTIONS => {
			'OPEN' => 9
		}
	},
	{#State 4
		DEFAULT => -1
	},
	{#State 5
		ACTIONS => {
			'' => 10
		}
	},
	{#State 6
		DEFAULT => -10
	},
	{#State 7
		ACTIONS => {
			'RE' => 11,
			'PROP' => 12
		},
		GOTOS => {
			'element' => 13
		}
	},
	{#State 8
		ACTIONS => {
			'TYPE' => 15
		},
		GOTOS => {
			'action' => 14
		}
	},
	{#State 9
		ACTIONS => {
			'RE' => 16,
			'NAME' => 17
		},
		GOTOS => {
			'tagname' => 18
		}
	},
	{#State 10
		DEFAULT => 0
	},
	{#State 11
		DEFAULT => -12
	},
	{#State 12
		ACTIONS => {
			'OPEN' => 19
		}
	},
	{#State 13
		DEFAULT => -3
	},
	{#State 14
		DEFAULT => -2
	},
	{#State 15
		ACTIONS => {
			'STR' => 20
		}
	},
	{#State 16
		DEFAULT => -9
	},
	{#State 17
		DEFAULT => -8
	},
	{#State 18
		ACTIONS => {
			'CLOSE' => 21
		}
	},
	{#State 19
		ACTIONS => {
			'NAME' => 22
		}
	},
	{#State 20
		DEFAULT => -13
	},
	{#State 21
		ACTIONS => {
			'DOT' => 23
		},
		DEFAULT => -5
	},
	{#State 22
		ACTIONS => {
			'CLOSE' => 24
		}
	},
	{#State 23
		ACTIONS => {
			'PROP' => 25,
			'CONTENT' => 26
		}
	},
	{#State 24
		DEFAULT => -11
	},
	{#State 25
		ACTIONS => {
			'OPEN' => 27
		}
	},
	{#State 26
		DEFAULT => -6
	},
	{#State 27
		ACTIONS => {
			'NAME' => 28
		}
	},
	{#State 28
		ACTIONS => {
			'CLOSE' => 29
		}
	},
	{#State 29
		DEFAULT => -7
	}
],
                                  yyrules  =>
[
	[#Rule 0
		 '$start', 2, undef
	],
	[#Rule 1
		 'function', 1,
sub
#line 31 "lib/HTML/Examiner/Parser.yp"
{ +{function=>$_[1]} }
	],
	[#Rule 2
		 'statement', 3,
sub
#line 34 "lib/HTML/Examiner/Parser.yp"
{ +{pattern=>$_[1],action=>$_[3]} }
	],
	[#Rule 3
		 'pattern', 3,
sub
#line 38 "lib/HTML/Examiner/Parser.yp"
{ +{target=>$_[1],verb=>$_[2],element=>$_[3]} }
	],
	[#Rule 4
		 'pattern', 1,
sub
#line 39 "lib/HTML/Examiner/Parser.yp"
{ +{target=>$_[1]} }
	],
	[#Rule 5
		 'target', 4,
sub
#line 42 "lib/HTML/Examiner/Parser.yp"
{ +{'tag'=>$_[3]} }
	],
	[#Rule 6
		 'target', 6,
sub
#line 43 "lib/HTML/Examiner/Parser.yp"
{ +{tag=>$_[3],content=>$_[6]} }
	],
	[#Rule 7
		 'target', 9,
sub
#line 44 "lib/HTML/Examiner/Parser.yp"
{ +{tag=>$_[3],prop=>$_[8]} }
	],
	[#Rule 8
		 'tagname', 1,
sub
#line 47 "lib/HTML/Examiner/Parser.yp"
{ +{'name'=>$_[1]} }
	],
	[#Rule 9
		 'tagname', 1,
sub
#line 48 "lib/HTML/Examiner/Parser.yp"
{ +{'re'=>$_[1]} }
	],
	[#Rule 10
		 'verb', 1,
sub
#line 51 "lib/HTML/Examiner/Parser.yp"
{ +{verb=>$_[1]} }
	],
	[#Rule 11
		 'element', 4,
sub
#line 54 "lib/HTML/Examiner/Parser.yp"
{ +{prop=>$_[3]} }
	],
	[#Rule 12
		 'element', 1,
sub
#line 55 "lib/HTML/Examiner/Parser.yp"
{ +{re=>$_[1]} }
	],
	[#Rule 13
		 'action', 2,
sub
#line 58 "lib/HTML/Examiner/Parser.yp"
{ +{type=>$_[1],msg=>$_[2]} }
	]
],
                                  @_);
    bless($self,$class);
}

#line 61 "lib/HTML/Examiner/Parser.yp"


=head2 lex

Function used to tokenize source code.

=cut

sub lex {
	my $self = shift;
#print "SOURCE ".$self->{'code'}."\n";

    for ($self->{'code'}) {
        s!^\s+!!;
        ($_ eq '')    and return ('',undef);

        s!^(tag)!!i    and return('TAG',$1);
        s!^(prop)!!i    and return('PROP',$1);
        s!^(error)!!i    and return('TYPE',$1);
        s!^(content)!!i    and return('CONTENT',$1);
        s!^(contains|misses|matches|mismatches)!!i    and return('VERB',$1);
        s!^(\=\>)!!    and return('ARROW',$1);
        s!^(\()!!    and return('OPEN',$1);
        s!^(\))!!    and return('CLOSE',$1);
        s!^(\.)!!    and return('DOT',$1);
        s!^(\w+)!!    and return('NAME',$1);
        #s!^(sub)!!    and return('SUB',$1);
        #s!^\{(.*)\}!!s    and print "|$1|\n" and return('CODE',$1);
        #s!^\{([^{}]*(\{[^{}]*\}[^{}]*)*)\}!!s and return('CODE',$1);

        s!^\"([^\"]+)\"!!    and return('STR',$1);
        s!^/([^/]*)/!!    and return('RE',$1);
    }
}

=head2 yyerror

Function used to report errors.

=cut

sub yyerror {
  if ($_[0]->YYCurtok) {
      printf STDERR ('Error: a "%s" (%s) was found where %s was expected'."\n",
         $_[0]->YYCurtok, $_[0]->YYCurval, $_[0]->YYExpect)
  }
  else { print  STDERR "Expecting one of ",join(", ",$_[0]->YYExpect),"\n";
  }
}

=head2 init_lex

Function used to initialize everything we need.

=cut

sub init_lex {
    my $self = shift;
    $self->{'code'} = shift;

    local $/;
    undef $/;
    #$File = <>
}

# vim: set filetype=perl

1;
